Improve per-vci thread-safety of Sched

rzambre opened the pull request:
The `per-VCI` critical section model is essentially a subset of the `per-object` model. It needs thread safety for each VCI, and thread-safety for the other objects such as requests, hooks, etc.

This PR improves the thread-safety of the `Nbc_progress_hook` which sits in `src/common/sched/mpidu_sched.c`. Currently, some of the operations are not thread-safe.

The PR first implements new naming rules for the existing functions so that it is easier to parse what kind of operations the functions do and what kind of objects they are acting upon. As a result, moving forward, it is straightforward to parse which functions need thread safety. 

Testing:
- [x] Rebase this on top of #3590 before running the tests.
- [x] Test with #3659 with all the nbc code since they rely on each other for completeness (which also includes the global lock for collectives in per-vci critical section).

rzambre commented:
@pavanbalaji could you review the first commit to see if the general approach is okay? After your feedback, I will do it the same way for the remaining hooks.

pavanbalaji commented on src/include/mpir_sched.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef MPIR_SCHED_H_INCLUDED
+#define MPIR_SCHED_H_INCLUDED
+
+typedef struct MPIDU_Sched_state MPIR_Sched_list_t;
```
The `MPIDU_` functions/datatypes should have similar names as the `MPIR_` versions.

pavanbalaji commented on src/mpi/coll/src/coll_impl.c:
```diff
@@ -956,7 +956,7 @@ int MPII_Coll_init(void)
 
     /* register non blocking collectives progress hook */
-    mpi_errno = MPID_Progress_register_hook(MPIDU_Sched_progress, &MPIR_Nbc_progress_hook_id);
+    mpi_errno = MPID_Progress_register_hook(MPIDU_Sched_list_progress, &MPIR_Nbc_progress_hook_id);
```
This is not a `sched_list` function.  It's not specific to a list or to a schedule.  How about we use the following naming:

`MPIR|MPIDU_Sched_element_*`
`MPIR|MPIDU_Sched_list_*`
`MPIR|MPIDU_Sched_global_*`

rzambre replied:
Just a note following our offline discussion. The list of all schedules is global and so `Sched_global` and `Sched_list` mean the same thing.

Using the following naming instead:
`MPIR|MPIDU_Sched_element_*`
`MPIR|MPIDU_Sched_list_*`

pavanbalaji commented on src/mpi/comm/commutil.c:
```diff
@@ -98,7 +98,7 @@ int MPII_Comm_init(MPIR_Comm * comm_p)
 
     /* abstractions bleed a bit here... :(*/
-    comm_p->next_sched_tag = MPIR_FIRST_NBC_TAG;
+    OPA_store_int(&comm_p->next_sched_tag, MPIR_FIRST_NBC_TAG);
```
That should be abstracted to only use atomics when threads are involved.

rzambre replied:
#3733 will address this abstraction.

pavanbalaji commented on src/mpid/ch3/channels/nemesis/include/mpid_nem_inline.h:
```diff
@@ -89,7 +89,7 @@ static inline void MPID_nem_mpich_send_seg (MPIR_Segment *segment, intptr_t *seg
      !MPIDI_CH3I_shm_active_send &&             \
      !MPIDI_CH3I_Sendq_head(MPIDI_CH3I_shm_sendq) &&       \
-     !MPIDU_Sched_are_pending() &&              \
+     !MPIDU_Sched_list_has_pending_sched() &&              \
```
Namespace should be `MPIDU_Sched_global_has_pending_sched`?

pavanbalaji commented on src/mpid/common/sched/mpidu_sched.c:
```diff
@@ -167,34 +170,42 @@ int MPIDU_Sched_next_tag(MPIR_Comm * comm_ptr, int *tag)
-    *tag = comm_ptr->next_sched_tag;
-    ++comm_ptr->next_sched_tag;
+    do {
+        my_sched_tag = OPA_load_int(&comm_ptr->next_sched_tag);
```
Atomic only when threads are involved.

rzambre replied:
#3733 will address this abstraction.

hzhou commented:
> This is WIP. Do not merge.

There is a label for that :)

hzhou commented:
I'll write down some comments I had from the last meeting. I haven't reviewed the actual commit of this PR, so it may not be directly related.

The whole idea of "context" is isolation. There is not much point of context if it is leaky, i.e. underlying components such as progress hooks can access states that crosses context. With that, it appears the only proper way to implement VCI is to create this global context pool, where each context contains a complete request pool, progress states, and underlying transport context. That means the necessary states/data for request pool and progress engines are all part of fields in this context structure, and requests and progress engines can only access within its context. With that isolation, we will only need a per VCI lock once. Specifically, by default, we will have a global context, and this context, other than the name, is a sibling to any new context created -- maybe a proper name would be default context. With that, we don't need any special handling for the global context, just the default context. Global progress can still be made by the entrance function -- `MPI_Wait` or `MPI_Waitall`, which happens before taking the context lock, thus can decide on the logic of polling which or each context, upon which, proper lock is taken. And once the context lock is taken, all functions never acts out-of-context.

raffenet commented on src/include/mpir_comm.h:
```diff
@@ -168,7 +168,7 @@ struct MPIR_Comm {
     struct MPII_Topo_ops *topo_fns;     /* Pointer to a table of functions
                                          * implementting the topology routines */
-    int next_sched_tag;         /* used by the NBC schedule code to allocate tags */
+    OPA_int_t next_sched_tag;   /* used by the NBC schedule code to allocate tags */
```
@rzambre and I discussed and I believe this doesn't need to be thread-safe. Collectives should only be invoked in order, one thread at a time. Any other usage would be an erroneous program.

rzambre commented:
> There is a label for that :)

@hzhou unfortunately, I don't have access to adding labels to the PRs and the issues :( And that's why I have been explicitly writing `WIP` and `Do not merge`. 

This PR is dealing with the thread safety of the hooks in the `per-VCI` model. The `per-VCI` model is essentially a subset of the `per-object` model. So, in the `per-VCI` model we need thread safety for each VCI, and thread-safety for the other objects such as requests, hooks, etc.

And regarding your point, let me think about our offline discussion :)

rzambre commented:
@pavanbalaji the latest commit has addressed your comments. I removed my commit for thread safety in hcoll since it needs some more work.

Here's a question regarding hcoll: who calls the hcoll runtime api (https://github.com/pmodels/mpich/blob/master/src/mpid/common/hcoll/hcoll_rte.c)? I see it is initialized during `hcoll_initialize()` but I don't see them being called in MPICH.

pavanbalaji commented:
@rzambre I should have been clearer.  The type renaming was supposed to be a prereq commit to the actual lock changes.  The type renaming (which I assume is simple scripting) is better reviewed by a compiler, while the lock changes are better reviewed by a human.  Can you do that?

The easiest way to do it will likely be to checkout `master`, do the type renaming (hopefully using a script), and then cherry-picking this commit on top of that branch.

pavanbalaji commented:
> Here's a question regarding hcoll: who calls the hcoll runtime api (https://github.com/pmodels/mpich/blob/master/src/mpid/common/hcoll/hcoll_rte.c)? I see it is initialized during `hcoll_initialize()` but I don't see them being called in MPICH.

That's really a question for @raffenet.  He's the one who has been poking around with `hcoll`.

rzambre commented:
@pavanbalaji yes, I ended up following your method. The latest version now has the commit split into two. Let me know if it looks okay.

pavanbalaji commented:
Looks good to me.

raffenet commented:
@rzambre `hcoll` calls the runtime API. MPICH sets up the functionality `hcoll` needs to initialize and perform collective operations.

rzambre commented:
I have rebased and pushed commits that maintain thread safety for the TSP code.

pavanbalaji commented:
This looks generally fine.  The fixes to the TSP code are orthogonal and should be moved to a different PR, so that can be more easily merged in.

Why are you using TSP_queue instead of TSP_sched_list, like you did with the sched patch?

I think locks should not simply protect full functions.  They should protect the parts that are accessing global data structures.  So, in the enqueue function, only the actual enqueue operation should be lock protected, not allocation of requests, etc.  (allocation of requests would be a different class that would have its own lock).

rzambre commented:
I have updated the title, description, and commits (for finer critical sections) of this PR. The WIP label can now be removed. I will continue the conversation in the new PR.

test:jenkins/ch3/most
test:jenkins/ch4/most

rzambre commented:
Rebased.

test:jenkins/ch3/most

rzambre commented:
test:jenkins/ch4/vci

raffenet commented:
test:jenkins/ch4/vci

raffenet commented:
test:jenkins/ch4/vci

raffenet commented:
test:jenkins/ch4/vci

rzambre commented:
[ ] Rebase this on top of #3590 before running the tests.
[ ] Remove the global lock on collectives before really testing the changes of this PR.

rzambre commented:
test:jenkins/ch4/vci

rzambre commented:
test:jenkins/ch4/vci

hzhou commented:
Obsoleted bu PR #4086
